### PR TITLE
introduction to footprints

### PR DESCRIPTION
**Pull request by ih64**
ready for review!

related to issue #131 on source detection, footprints, heavy footprints.

so far i have some tutorial level introduction to footprints and the methods they support. Some ideas on including the NoiseGenerator to replace footprints with noise, and add them back one at a time, as this is what happens during measurement. 

How should peaks, deblending, and the difference between single visit and coadd visits be included in the notebook?

### CODE REVIEW

**Comment by ih64** [comment]
alrighty, ready for some feedback from a review!

**Comment by drphilmarshall** [comment]
Thanks for a nice walkthrough earlier, @ih64 ! Good reviewers for this notebook could be: @kadrlica (but he gets asked a lot), @jeffcarlin , @fred3m . Feel free to request a review from one of them, or anyone else. And of course, other clubmates, you can volunteer to review notebooks any time!

**Comment by kadrlica** [comment]
I'm happy to review this notebook. I'll assign myself, and @ih64 should let me know when it's ready.

**Comment by ih64** [comment]
Hi @kadrlica, thank you! Is it ok if I ping you around Wednesday next week? There's one feature I want to add, and I will put it in by then

**Comment by kadrlica** [comment]
Absolutely, just let me know when I should start.

File: SourceDetection/Footprints.ipynb [hunk 1]
```diff
@@ -0,0 +1,1207 @@
...
+    "for i in src_ids:\n",
+    "    noiseReplacer.insertSource(i)\n",
+    "    data = calexp.getMaskedImage()[2600:2800,2900:3100].getImage().array\n",
+    "    im = plt.imshow(data, origin='lower', animated=True, vmin=-100, vmax=100)\n",
+    "    ims.append([im])\n",
+    "    noiseReplacer.removeSource(i)\n",
+    "\n",
+    "HTML(animation.ArtistAnimation(fig, ims, interval=1000, blit=True,\n",
+    "                                repeat_delay=1000).to_jshtml())\n",
+    "# takes a minute to get going, "
```
Comment (kadrlica) [inline SourceDetection/Footprints.ipynb]: It looks like this is outputting a spurious second image. I'm not sure how to get rid of this (maybe an artifact of imshow)?

Comment (kadrlica) [reply SourceDetection/Footprints.ipynb]: This is what I see...

image

File: SourceDetection/Footprints.ipynb [hunk 2]
```diff
@@ -0,0 +1,1214 @@
...
+   "source": [
+    "# first get the gaussian fwhm of the PSF \n",
+    "psf = calexp.getPsf()\n",
+    "sigma = psf.computeShape().getDeterminantRadius()\n",
+    "\n",
+    "# print how many pixels belong to the footprint before we dilate it\n",
+    "print('before dilating the footprint has an area of {}'.format(fps[0].getArea()))\n",
+    "\n",
+    "# now we will grow the footprint by 2 x sigma. We need to pass an int\n",
+    "fps[0].dilate(int(2*sigma))\n",
```
[the remark this answers is not in the document]
Comment (ih64) [reply SourceDetection/Footprints.ipynb]: I looked at Jim's paper and the doxygen on dilation/erosion, and you are right, it is like the scipy operation. I added a few words into this description to clear that up

File: SourceDetection/Footprints.ipynb [hunk 3]
```diff
@@ -0,0 +1,1214 @@
...
+    "\n",
+    "# print how many pixels belong to the footprint ater dilation\n",
+    "print('after eroding the footprint has an area of {}'.format(fps[0].getArea()))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To give us some more intuition on how dilation effects the footprints, we will show cut outs of several footprints before they have been dilated (top row) and after they have been dilated. To exagerate the effect, we will use a large scale length in the dilation."
```
[the remark this answers is not in the document]
Comment (ih64) [reply SourceDetection/Footprints.ipynb]: third row added

File: SourceDetection/Footprints.ipynb [hunk 4]
```diff
@@ -0,0 +1,1214 @@
...
+   "metadata": {},
+   "source": [
+    "## Erosion and Dilation on Footprints"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can grow or shrink the PSF using the dilate() and erode() methods. You may imagine doing this if you want to ensure you are getting all the flux associated with the detected source. Using these methods requires that we provide a scale by which to grow or shrink the PSF. Typically, this scale is the $\\sigma$ of the gaussian PSF scaled by a user defined factor. Let us demonstrate this below"
```
[the remark this answers is not in the document]
Comment (ih64) [reply SourceDetection/Footprints.ipynb]: yes, my mistake, it is by the psf

File: SourceDetection/Footprints.ipynb [hunk 5]
```diff
@@ -0,0 +1,1214 @@
...
+   "outputs": [],
+   "source": [
+    "calexp.getMask().getMaskPlaneDict()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The values are the exponent of the bitmask. So Pixels only marked detected will be 2^5 = 32. Pixels that are both on the edge and detected will be 2^5 + 2^4 = 48. Now we will visualize this in a simmilar manner to the afwdisplay exercise we did before, only now we are *only* using data for the footprint because we are using the span."
```
[the remark this answers is not in the document]
Comment (ih64) [reply SourceDetection/Footprints.ipynb]: thanks for the catch, changed to imshow

File: SourceDetection/Footprints.ipynb [hunk 6]
```diff
@@ -0,0 +1,1214 @@
...
+    "Now we have drilled our way down to getting our hands on footprints. Lets discuss some of their methods and attributes. To begin, we will concentrate on Footprint geometry and location."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Get the region-the pixel coordinates that describes the chip this footprint belongs too\n",
```
[the remark this answers is not in the document]
Comment (ih64) [reply SourceDetection/Footprints.ipynb]: added some clarification, region is the whole CCD, bbox is always inside the ccd

File: SourceDetection/Footprints.ipynb (hunk 1 above)
Comment (ih64) [reply SourceDetection/Footprints.ipynb]: should be fixed now

**Pull request merged**